syntax = "proto3";
package backup_protocol;

service Replica{
    rpc Read(ReadDeleteRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (WriteResponse);
    rpc Delete(ReadDeleteRequest) returns (Response);
    rpc GetAllData(ReadDeleteRequest) returns (AllData);
}

service RegistryServer{
    rpc RegisterReplica(ServerMessage) returns (ServerMessage);
    rpc GetReplicas(RequestType) returns (ServerListResponse);
}

message ServerMessage {
    string uuid = 1;
    string address = 2;
}

message ServerListResponse {
    repeated ServerMessage serverList = 1;
}

message WriteRequest {
    string name = 1;
    string content = 2;
    string uuid = 3;
}

message WriteResponse {
    Status status = 1;
    string uuid  = 2;
    string version = 3;
}

message ReadDeleteRequest{
    string uuid=1;
}

message AllData{
    repeated ReadResponse readResponse = 1;
}

message ReadResponse{
    Status status=1;
    string name=2;
    string content=3;
    string version=4;
}

message Response{
    Status response=1;
    string reason=2;
}

message FileDetail{
    string file_name=1;
    string version=2;
}

message FileList{
    repeated FileDetail fileList = 1;
}

// new added for quorun_protocol
message RequestType{
    string type = 1;
}

enum Status {
    SUCCESS = 0;
    FAIL = 1;
}